pkg/kubectl/resize.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+)
+
+// ResizePrecondition describes a condition that must be true for the resize to take place
+// If CurrentSize == -1, it is ignored.
+// If CurrentResourceVersion is the empty string, it is ignored.
+// Otherwise they must equal the values in the replication controller for it to be valid.
+type ResizePrecondition struct {
+	Size            int
+	ResourceVersion string
+}
+
+type PreconditionError struct {
+	Precondition  string
+	ExpectedValue string
+	ActualValue   string
+}
+
+func (pe *PreconditionError) Error() string {
+	return fmt.Sprintf("Expected %s to be %s, was %s", pe.Precondition, pe.ExpectedValue, pe.ActualValue)
+}
+
+// Validate ensures that the preconditions match.  Returns nil if they are valid, an error otherwise
+func (precondition *ResizePrecondition) Validate(controller *api.ReplicationController) error {
+	if precondition.Size != -1 && controller.Spec.Replicas != precondition.Size {
+		return &PreconditionError{"replicas", strconv.Itoa(precondition.Size), strconv.Itoa(controller.Spec.Replicas)}
+	}
+	if precondition.ResourceVersion != "" && controller.ResourceVersion != precondition.ResourceVersion {
```

Comment:
Due to races, this operation could still fail due to a resource version conflict on the server. I don't think 2 different types of errors should be reported. 

Both of these preconditions should eventually be implemented on the server. Right now, the resource version precondition could be just checked by the server. We could potentially retry the size precondition if the operation failed due to resource version conflict.